[ipcamera] Discovery crashes immediately

<!-- Provide a general summary of the issue in the *Title* above -->
<!-- If the issue is related to a binding, please include its short name in -->
<!-- square brackets in the title - Example: "[astro] My issue..." -->

<!-- Important: Please contact the openHAB community forum for questions or -->
<!-- for configuration and usage guidance: https://community.openhab.org -->

<!-- Feel free to delete any comment lines in the template (starting with "<!--") -->

## Expected Behavior
<!-- If you're describing a bug, tell us what should happen -->
<!-- If you're suggesting a change/improvement, tell us how it should work -->
The discovery process finds IP cameras on my network.
When starting a manual discovery (with the Scan button), the progress bar on the `/settings/things/add/ipcamera` page gets filled up.

## Current Behavior
<!-- If describing a bug, tell us what happens instead of the expected behavior -->
<!-- Include related log information (preferably debug level) and related configs -->
<!-- Use a file attachment for log and config information longer than a few lines -->
<!-- Enclose multi-line log/code snippets with ``` on new lines for proper formatting -->
<!-- If suggesting a change/improvement, explain the difference from current behavior -->
<!-- For improvements, discuss at community.openhab.org first and include link to topic -->
The discovery process does not find any IP cameras on my network.
When starting a manual discovery (with the Scan button), the progress bar on the `/settings/things/add/ipcamera` page disappears in a split second.

I have uploaded logs [here](https://gist.github.com/mpeter50/a17ccf9b8f4a99b51e01af871a4856fe), with default log levels plus TRACE level for `org.jupnp` and `org.openhab.binding.ipcamera`.

The openHAB instance does not have special configuration, it was just installed, and the first action on it being done was adding this binding.

There are several other people on the openHAB forum who have this issue, the earliest report I have found being from February of 2021, but so far no solution has been reported: https://community.openhab.org/t/ipcamera-discovery-error/116790

## Steps to Reproduce (for Bugs)
<!-- Provide a link to a live example, or an unambiguous set of steps to -->
<!-- reproduce this bug. Include code to reproduce, if relevant -->
1. Install openHAB in a docker container according to the first compose file example under "Running from compose-file.yml" [here](https://hub.docker.com/r/openhab/openhab/)
2. Open the web interface
3. Install the binding-ipcamera binding
4. Start a scan manually

## Context
<!-- How has this issue affected you? What are you trying to accomplish? -->
<!-- Providing context helps us come up with a solution that is most useful in the real world -->

I am trying to automatically discover my ONVIF compatible IP cameras.

## Your Environment
<!-- Include as many relevant details about the environment you experienced the bug in -->
* Version used:
  * openHAB container image: openhab/openhab:3.4.2
  * binding-ipcamera version: 3.4.2 
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...):
  * Desktop
  * openSUSE Leap 15.4 as Docker host
  * Docker version: 20.10.17-ce, build a89b84221c85

I have seen the below error before and I believe it occurs because 1 scan has not fully finished before the 2nd scan is triggered. I did a restart and it was fixed.

```
2023-02-25 13:12:42.692 [ERROR] [nternal.DiscoveryServiceRegistryImpl] - Cannot trigger scan for thing types '[ipcamera:generic, ipcamera:dahua, ipcamera:instar, ipcamera:amcrest, ipcamera:hikvision, ipcamera:onvif, ipcamera:foscam, ipcamera:doorbird]' on 'IpCameraDiscoveryService'!
io.netty.channel.ChannelPipelineException: org.openhab.binding.ipcamera.internal.onvif.OnvifDiscovery$2 is not a @Sharable handler, so can't be added or removed multiple times.
	at io.netty.channel.DefaultChannelPipeline.checkMultiplicity(DefaultChannelPipeline.java:600) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:202) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:381) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:370) ~[?:?]
	at io.netty.bootstrap.Bootstrap.init(Bootstrap.java:262) ~[?:?]
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:311) ~[?:?]
	at io.netty.bootstrap.AbstractBootstrap.doBind(AbstractBootstrap.java:272) ~[?:?]
	at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:268) ~[?:?]
	at org.openhab.binding.ipcamera.internal.onvif.OnvifDiscovery.discoverCameras(OnvifDiscovery.java:220) ~[?:?]
```

I would need to look at this to see what is involved if anything to make the Netty handler be able to be tagged @Sharable. Either stop having multiple scans at the same time, or make it possible to do parallel scans.

--- a/bundles/org.openhab.binding.ipcamera/README.md
+++ b/bundles/org.openhab.binding.ipcamera/README.md
@@ -133,10 +133,11 @@ Thing ipcamera:hikvision:West "West Camera"
 
 ## Discovery
 
-The discovery feature of openHAB can be used to find and setup any ONVIF cameras.
+The discovery feature of openHAB can be used to find and setup ONVIF cameras.
 This method should be preferred as it will discover the cameras IP, ports and URLs for you, making the setup much easier.
 The binding needs to use UDP port 3702 to discover the cameras with, so this port needs to be unblocked by your firewall or add the camera manually if the camera is not auto found.
 To use the discovery, just press the `+` icon located in the Inbox, then select the IpCamera binding from the list of installed bindings.
+The binding will only search using openHAB's currently selected primary network address, see <https://www.openhab.org/docs/settings/>.
 If your camera is not found after a few searches, it may not be ONVIF and in this case you will need to manually add the camera via the UI.
 Cameras that are not ONVIF should be added as a `generic` thing type and you will need to provide the URLs manually.
 

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/IpCameraDiscoveryService.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/IpCameraDiscoveryService.java
@@ -22,9 +22,12 @@ import org.openhab.core.config.discovery.AbstractDiscoveryService;
 import org.openhab.core.config.discovery.DiscoveryResult;
 import org.openhab.core.config.discovery.DiscoveryResultBuilder;
 import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.net.NetworkAddressService;
 import org.openhab.core.thing.ThingTypeUID;
 import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
 import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -39,9 +42,12 @@ import org.slf4j.LoggerFactory;
 public class IpCameraDiscoveryService extends AbstractDiscoveryService {
 
     private final Logger logger = LoggerFactory.getLogger(IpCameraDiscoveryService.class);
+    private final NetworkAddressService networkAddressService;
 
-    public IpCameraDiscoveryService() {
+    @Activate
+    public IpCameraDiscoveryService(@Reference NetworkAddressService networkAddressService) {
         super(SUPPORTED_THING_TYPES, 0, false);
+        this.networkAddressService = networkAddressService;
     }
 
     @Override
@@ -65,7 +71,7 @@ public class IpCameraDiscoveryService extends AbstractDiscoveryService {
     @Override
     protected void startScan() {
         removeOlderResults(getTimestampOfLastScan());
-        OnvifDiscovery onvifDiscovery = new OnvifDiscovery(this);
+        OnvifDiscovery onvifDiscovery = new OnvifDiscovery(networkAddressService, this);
         try {
             onvifDiscovery.discoverCameras();
         } catch (UnknownHostException | InterruptedException e) {

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
@@ -144,8 +144,18 @@ public class OnvifDiscovery {
             if (!xAddr.isEmpty()) {
                 searchReply(xAddr, xml);
             } else if (xml.contains("onvif")) {
-                logger.info("Possible ONVIF camera found at:{}", packet.sender().getHostString());
-                ipCameraDiscoveryService.newCameraFound("onvif", packet.sender().getHostString(), 80);
+                String brand;
+                try {
+                    brand = getBrandFromLoginPage(packet.sender().getHostString());
+                } catch (IOException e) {
+                    brand = "onvif";
+                }
+                logger.info("Possible {} camera found at:{}", brand, packet.sender().getHostString());
+                if (brand.equals("reolink")) {
+                    ipCameraDiscoveryService.newCameraFound(brand, packet.sender().getHostString(), 80);
+                } else {
+                    ipCameraDiscoveryService.newCameraFound(brand, packet.sender().getHostString(), 80);
+                }
             }
         }
     }
@@ -155,14 +165,16 @@ public class OnvifDiscovery {
             return "dahua";
         } else if (response.toLowerCase().contains("dahua")) {
             return "dahua";
+        } else if (response.toLowerCase().contains("doorbird")) {
+            return "doorbird";
         } else if (response.toLowerCase().contains("foscam")) {
             return "foscam";
         } else if (response.toLowerCase().contains("hikvision")) {
             return "hikvision";
         } else if (response.toLowerCase().contains("instar")) {
             return "instar";
-        } else if (response.toLowerCase().contains("doorbird")) {
-            return "doorbird";
+        } else if (response.toLowerCase().contains("reolink")) {
+            return "reolink";
         } else if (response.toLowerCase().contains("ipc-")) {
             return "dahua";
         } else if (response.toLowerCase().contains("dh-sd")) {

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
@@ -90,8 +90,9 @@ public class OnvifDiscovery {
                             && inetAddress.isSiteLocalAddress()) {
                         if (inetAddress.getHostAddress().equals(primaryHostAddress)) {
                             results.add(networkInterface);
+                            logger.info("Scanning network {} for any ONVIF cameras", primaryHostAddress);
                         } else {
-                            logger.debug("Skipping network {} as it was not selected as openHAB's primary network",
+                            logger.debug("Skipping network {} as it was not selected as openHAB's 'Primary Address'",
                                     inetAddress.getHostAddress());
                         }
                     } else {
@@ -210,7 +211,8 @@ public class OnvifDiscovery {
     public void discoverCameras() throws UnknownHostException, InterruptedException {
         List<NetworkInterface> nics = getLocalNICs();
         if (nics == null || nics.isEmpty()) {
-            logger.debug("No valid networks detected to use for camera discovery");
+            logger.warn(
+                    "No 'Primary Address' selected to use for camera discovery. Check openHAB's Network Settings page to select a valid Primary Address.");
             return;
         }
         NetworkInterface networkInterface = nics.get(0);

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
@@ -151,8 +151,8 @@ public class OnvifDiscovery {
                     brand = "onvif";
                 }
                 logger.info("Possible {} camera found at:{}", brand, packet.sender().getHostString());
-                if (brand.equals("reolink")) {
-                    ipCameraDiscoveryService.newCameraFound(brand, packet.sender().getHostString(), 80);
+                if ("reolink".equals(brand)) {
+                    ipCameraDiscoveryService.newCameraFound(brand, packet.sender().getHostString(), 8000);
                 } else {
                     ipCameraDiscoveryService.newCameraFound(brand, packet.sender().getHostString(), 80);
                 }
@@ -179,8 +179,6 @@ public class OnvifDiscovery {
             return "dahua";
         } else if (response.toLowerCase().contains("dh-sd")) {
             return "dahua";
-        } else if (response.toLowerCase().contains("reolink")) {
-            return "reolink";
         }
         return "onvif";
     }

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
@@ -35,6 +35,7 @@ import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
 import org.openhab.binding.ipcamera.internal.Helper;
 import org.openhab.binding.ipcamera.internal.IpCameraDiscoveryService;
+import org.openhab.core.net.NetworkAddressService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -62,16 +63,21 @@ import io.netty.util.concurrent.GlobalEventExecutor;
  */
 
 @NonNullByDefault
+@io.netty.channel.ChannelHandler.Sharable
 public class OnvifDiscovery {
     private IpCameraDiscoveryService ipCameraDiscoveryService;
     private final Logger logger = LoggerFactory.getLogger(OnvifDiscovery.class);
+    private final NetworkAddressService networkAddressService;
     public ArrayList<DatagramPacket> listOfReplys = new ArrayList<DatagramPacket>(10);
 
-    public OnvifDiscovery(IpCameraDiscoveryService ipCameraDiscoveryService) {
+    public OnvifDiscovery(NetworkAddressService networkAddressService,
+            IpCameraDiscoveryService ipCameraDiscoveryService) {
         this.ipCameraDiscoveryService = ipCameraDiscoveryService;
+        this.networkAddressService = networkAddressService;
     }
 
     public @Nullable List<NetworkInterface> getLocalNICs() {
+        String primaryHostAddress = networkAddressService.getPrimaryIpv4HostAddress();
         List<NetworkInterface> results = new ArrayList<>(2);
         try {
             for (Enumeration<NetworkInterface> enumNetworks = NetworkInterface.getNetworkInterfaces(); enumNetworks
@@ -82,7 +88,14 @@ public class OnvifDiscovery {
                     InetAddress inetAddress = enumIpAddr.nextElement();
                     if (!inetAddress.isLoopbackAddress() && inetAddress.getHostAddress().toString().length() < 18
                             && inetAddress.isSiteLocalAddress()) {
-                        results.add(networkInterface);
+                        if (inetAddress.getHostAddress().equals(primaryHostAddress)) {
+                            results.add(networkInterface);
+                        } else {
+                            logger.debug("Skipping network {} as it was not selected as openHAB's primary network",
+                                    inetAddress.getHostAddress());
+                        }
+                    } else {
+                        logger.debug("Skipping network {} as it was not site local", inetAddress.getHostAddress());
                     }
                 }
             }
@@ -197,6 +210,7 @@ public class OnvifDiscovery {
     public void discoverCameras() throws UnknownHostException, InterruptedException {
         List<NetworkInterface> nics = getLocalNICs();
         if (nics == null || nics.isEmpty()) {
+            logger.debug("No valid networks detected to use for camera discovery");
             return;
         }
         NetworkInterface networkInterface = nics.get(0);

--- a/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
+++ b/bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifDiscovery.java
@@ -90,7 +90,7 @@ public class OnvifDiscovery {
                             && inetAddress.isSiteLocalAddress()) {
                         if (inetAddress.getHostAddress().equals(primaryHostAddress)) {
                             results.add(networkInterface);
-                            logger.info("Scanning network {} for any ONVIF cameras", primaryHostAddress);
+                            logger.debug("Scanning network {} for any ONVIF cameras", primaryHostAddress);
                         } else {
                             logger.debug("Skipping network {} as it was not selected as openHAB's 'Primary Address'",
                                     inetAddress.getHostAddress());

